refactor: simplify watchmap traversal

baszalmstra opened the pull request:
This PR cleans up the code by adding a new type for traversing and modifying the watchmap. The propagation loop contained a number of bookkeeping locals which distract from the actual propagation algorithm. There were also methods to manipulate the watches and the watchmap in different places, they are now all consolidated in the `WatchMapCursor` struct. 

Using the cursor is very simple: `WatchMap::cursor` is used to create a cursor to iterate the linked list of watches. `WatchMapCursor::next` moves the cursor to the next clause watching a certain literal. `WatchMapCursor::update` modifies the current clause/watch to watch a different literal. This method performs all the bookkeeping of the linked lists to ensure consistency.

Hopefully this makes the code easier to understand.

tdejager commented on src/internal/id.rs:
```diff
@@ -114,6 +114,7 @@ impl ArenaId for ClauseId {
         Self(unsafe { NonZeroU32::new_unchecked((x + 1).try_into().expect("clause id too big")) })
     }
 
+    #[inline]
```
Wouldn't this be inlined automatically? It seems trivial if it could and otherwise it might not benefit?

baszalmstra replied:
Not needed

```suggestion
```

tdejager commented on src/solver/clause.rs:
```diff
@@ -503,40 +481,33 @@ impl ClauseState {
 #[repr(transparent)]
 #[derive(Debug, Copy, Clone, Eq, PartialEq)]
-pub(crate) struct Literal(u32);
+pub(crate) struct Literal(NonZeroU32);
```
Dit we mention the autmatic optimization for size that rust does here somewhere?

baszalmstra replied:
I added a comment, and a test!

tdejager commented on src/solver/clause.rs:
```diff
@@ -503,40 +481,33 @@ impl ClauseState {
 
 impl ArenaId for Literal {
     fn from_usize(x: usize) -> Self {
-        debug_assert!(x <= u32::MAX as usize, "watched literal id too big");
```
Was this assert never hit?

tdejager replied:
Ah is see its the stament below now.

tdejager commented on src/solver/clause.rs:
```diff
@@ -503,40 +481,33 @@ impl ClauseState {
+        unsafe { Literal(NonZeroU32::new_unchecked(idx)) }
     }
 
     fn to_usize(self) -> usize {
```
There was a previous `#[inline]` in this PR, if we need it there we need it here I think?

baszalmstra replied:
I removed all `#[inline]` statements.

tdejager commented on src/solver/clause.rs:
```diff
@@ -321,14 +322,15 @@ impl Clause {
-pub(crate) struct ClauseState {
-    // The ids of the literals this clause is watching
-    pub watched_literals: [Literal; 2],
+pub(crate) struct ClauseWatches {
```
Wouldn't `WatchedLiterals` make more sense, also reading the description?

baszalmstra replied:
Yes much better, changed it!

tdejager replied:
Glad to be of service :)

baszalmstra commented:
Mm the benchmarks are not favorable:

```
main_timings.csv: 1000 records
timings.csv: 1000 records

Summary for main_timings.csv:
- Average Solve Time: 0.83 seconds (mean of all durations)
- Median Solve Time: 0.35 seconds (middle value when sorted)
- Standard Deviation: 2.95 seconds (spread of durations around the mean)
- Minimum Solve Time: 0.00 seconds (shortest solve duration)
- Maximum Solve Time: 50.00 seconds (longest solve duration, capped at 50s)
- 25th Percentile: 0.11 seconds (25% of solves were faster than this)
- 75th Percentile: 0.89 seconds (75% of solves were faster than this)
- Number of Outliers: 3 (solves capped at 50s)

Summary for timings.csv:
- Average Solve Time: 0.94 seconds (mean of all durations)
- Median Solve Time: 0.40 seconds (middle value when sorted)
- Standard Deviation: 3.06 seconds (spread of durations around the mean)
- Minimum Solve Time: 0.00 seconds (shortest solve duration)
- Maximum Solve Time: 50.00 seconds (longest solve duration, capped at 50s)
- 25th Percentile: 0.12 seconds (25% of solves were faster than this)
- 75th Percentile: 0.98 seconds (75% of solves were faster than this)
- Number of Outliers: 3 (solves capped at 50s)

Comparison between the datasets:
- Average Solve Time: 'timings.csv' was 0.88 times faster than 'main_timings.csv'
- Median Solve Time: 'timings.csv' was 0.88 times faster than 'main_timings.csv'
- 25th Percentile: 'timings.csv' was 0.93 times faster than 'main_timings.csv'
- 75th Percentile: 'timings.csv' was 0.91 times faster than 'main_timings.csv'
- Outliers: 'timings.csv' had 0 fewer solves capped at 50s
```

![Untitled](https://github.com/user-attachments/assets/cb35a6e4-4602-4987-8ce7-9e9437b119fe)

baszalmstra commented:
I simplified the code even further and added a little bit of unsafe magic and the benchmarks look much better now:

```
main_timings.csv: 1000 records
timings.csv: 1000 records

Summary for main_timings.csv:
- Average Solve Time: 0.83 seconds (mean of all durations)
- Median Solve Time: 0.35 seconds (middle value when sorted)
- Standard Deviation: 2.95 seconds (spread of durations around the mean)
- Minimum Solve Time: 0.00 seconds (shortest solve duration)
- Maximum Solve Time: 50.00 seconds (longest solve duration, capped at 50s)
- 25th Percentile: 0.11 seconds (25% of solves were faster than this)
- 75th Percentile: 0.89 seconds (75% of solves were faster than this)
- Number of Outliers: 3 (solves capped at 50s)

Summary for timings.csv:
- Average Solve Time: 0.86 seconds (mean of all durations)
- Median Solve Time: 0.37 seconds (middle value when sorted)
- Standard Deviation: 3.00 seconds (spread of durations around the mean)
- Minimum Solve Time: 0.00 seconds (shortest solve duration)
- Maximum Solve Time: 50.00 seconds (longest solve duration, capped at 50s)
- 25th Percentile: 0.11 seconds (25% of solves were faster than this)
- 75th Percentile: 0.89 seconds (75% of solves were faster than this)
- Number of Outliers: 3 (solves capped at 50s)

Comparison between the datasets:
- Average Solve Time: 'timings.csv' was 0.96 times faster than 'main_timings.csv'
- Median Solve Time: 'timings.csv' was 0.95 times faster than 'main_timings.csv'
- 25th Percentile: 'timings.csv' was 1.01 times faster than 'main_timings.csv'
- 75th Percentile: 'timings.csv' was 0.99 times faster than 'main_timings.csv'
- Outliers: 'timings.csv' had 0 fewer solves capped at 50s
```

![Untitled](https://github.com/user-attachments/assets/ed39c702-d692-4020-b69f-ed7a9cff61e5)

I feel this is good enough to merge.
